packages/react-components/react-aria/src/hooks/useARIAButton.ts

```diff
@@ -75,37 +93,55 @@ export const useARIAButton: ResolveShorthandFunction<ARIAButtonSlotProps> = (sho
     }
   });
 
-  if (shorthandProps) {
-    // If a <button> tag is to be rendered we just need to set disabled and aria-disabled correctly
-    if (shorthandProps.as === 'button' || shorthandProps.as === undefined) {
-      shorthandProps.disabled = disabled && !disabledFocusable;
-      shorthandProps['aria-disabled'] = disabledFocusable ? true : undefined;
-
-      // Undefine events if disabledFocusable is passed in
-      if (disabledFocusable) {
-        shorthandProps.onClick = undefined;
-        shorthandProps.onKeyDown = undefined;
-        shorthandProps.onKeyUp = undefined;
-      }
-    }
+  // If a <button> tag is to be rendered we just need to set disabled and aria-disabled correctly
+  if (type === 'button') {
+    return {
+      ...rest,
+      role,
+      tabIndex,
+      disabled: disabled && !disabledFocusable,
+      'aria-disabled': disabledFocusable ? true : undefined,
+      // Undefine event handlers if disabledFocusable is passed in
+      onClick: disabledFocusable ? undefined : onClick,
+      onKeyUp: disabledFocusable ? undefined : onKeyUp,
+      onKeyDown: disabledFocusable ? undefined : onKeyDown,
+    } as Props;
+  }
 
-    // If an <a> tag is to be rendered we have to remove disabled and type, and set aria-disabled, role and tabIndex.
-    else {
-      delete shorthandProps.disabled;
-      shorthandProps['aria-disabled'] = disabled || disabledFocusable;
-      (shorthandProps as JSX.IntrinsicElements['a']).href = disabled
+  // If an <a> or <div> tag is to be rendered we have to remove disabled and type,
+  // and set aria-disabled, role and tabIndex.
+  else {
+    const nextProps = {
+      ...rest,
+      role: role ?? 'button',
+      onClick: onClickHandler,
+      onKeyUp: onKeyupHandler,
+      onKeyDown: onKeyDownHandler,
+      'aria-disabled': disabled || disabledFocusable,
+      tabIndex: disabled && !disabledFocusable ? undefined : tabIndex ?? 0,
+    } as Props;
+
+    if (type === 'a') {
+      (nextProps as JSX.IntrinsicElements['a']).href = disabled
         ? undefined
-        : (shorthandProps as JSX.IntrinsicElements['a']).href;
-      shorthandProps.onClick = onClickHandler;
-      shorthandProps.onKeyDown = onKeyDownHandler;
-      shorthandProps.onKeyUp = onKeyupHandler;
-      shorthandProps.role = shorthandProps.role ?? 'button';
-      shorthandProps.tabIndex = disabled && !disabledFocusable ? undefined : tabIndex ?? 0;
+        : (props as JSX.IntrinsicElements['a']).href;
```

Comment:
is it necessary to pass the href down again ? shouldn't it be already in the `...rest` destructure ?

Reply:
Yeah, but it should be removed if `disabled` is true. I'll update it!